src/lib/snapp.ts

```diff
@@ -292,31 +373,59 @@ export class SmartContract {
     // TODO: think about how address should be passed in
     // if address is not provided, create a random one
     address ??= PrivateKey.random().toPublicKey();
-    let dummyInstance = new this(address);
-    let inductiveRules = (this._methods ?? []).map(
-      ({ methodName, witnessArgs }) => {
-        function main(transactionHash: Field) {
-          Mina.setCurrentTransaction({
-            sender: PrivateKey.random(), // TODO
-            parties: [],
-            nextPartyIndex: 0,
-            protocolState: ProtocolStatePredicate.ignoreAll(),
-          });
-          let witnesses = witnessArgs.map(emptyWitness); // TODO is this correct?
-          (dummyInstance[methodName] as any)(...witnesses); // call method in a way that `this` is defined
-          let hash = computeTransactionHash(Mina.currentTransaction);
-          console.log({ transactionHash, hash });
-          hash.assertEquals(transactionHash);
-          Mina.setCurrentTransaction(undefined);
-        }
-        return [0, methodName, main];
-      }
+    let instance = new this(address);
+
+    let rules = (this._methods ?? []).map(({ methodName, witnessArgs }) =>
+      picklesRuleFromFunction(
+        methodName,
+        (...args: unknown[]) => (instance[methodName] as any)(...args),
+        witnessArgs
+      )
+    );
+
+    let output = withContext(
+      {
+        self: Party.defaultParty(address),
+        protocolState: ProtocolStatePredicate.ignoreAll(),
+      },
+      () => picklesCompile(rules)
     );
-    let output = picklesCompile(inductiveRules);
-    console.log('output', output);
     return output;
   }
 
+  async prove(provers: any[], methodName: keyof this, args: unknown[]) {
+    let SnappClass = this.constructor as never as typeof SmartContract;
+    let i = SnappClass._methods!.findIndex((m) => m.methodName === methodName);
+    if (!(i + 1)) throw Error(`Method ${methodName} not found!`);
+    let ctx = {
+      self: Party.defaultParty(this.address),
+      protocolState: ProtocolStatePredicate.ignoreAll(),
+    };
+    // console.log(
+    //   'prove: running method directly in prover mode to get statement'
+    // );
+    let statement = Circuit.runAndCheckSync(() => {
+      mainContext = ctx;
+      (this[methodName] as any)(...args);
```

Comment:
can we use `withContext` (and the version without the global variable)

Reply:
done https://github.com/o1-labs/snarkyjs/commit/84a014996eaadf157dd7fc890b5e36fc51993f74